### PR TITLE
feat(skyrim-platform): implement missing Papyrus events

### PR DESCRIPTION
**Pull request by shevernitskiy**
Covered events can be found in this [table](https://docs.google.com/spreadsheets/d/1q-vZN34kgSxP7shJTZQ39vLJwqUJZUdf1pI5G54zA8Y/edit#gid=1825659776)

### CODE REVIEW

File: skyrim-platform/src/platform_se/skyrim_platform/GameEventSinks.cpp [hunk 1]
```diff
@@ -84,6 +85,36 @@ RE::BSEventNotifyControl GameEventSinks::ProcessEvent(
...
+    return RE::BSEventNotifyControl::kContinue;
+  }
+
+  auto waitStart = converted->waitStartTime;
+  auto waitEnd = converted->desiredWaitEndTime;
+
+  SkyrimPlatform::GetSingleton().AddUpdateTask([waitStart, waitEnd] {
+    auto obj = JsValue::Object();
+
+    if (waitStart < 0 || waitEnd < 0) {
```
Comment (Pospelove) [inline skyrim-platform/src/platform_se/skyrim_platform/GameEventSinks.cpp:105]: What's the meaning of this condition? Seems to be always false.
Please add a comment if this check is needed or remove.

File: skyrim-platform/src/platform_se/skyrim_platform/GameEventSinks.cpp [hunk 2]
```diff
@@ -84,6 +85,36 @@ RE::BSEventNotifyControl GameEventSinks::ProcessEvent(
...
+      return;
+    }
+
+    obj.SetProperty("startTime", JsValue::Double(waitStart));
+    obj.SetProperty("desiredEndTime", JsValue::Double(waitEnd));
```
Comment (Pospelove) [inline skyrim-platform/src/platform_se/skyrim_platform/GameEventSinks.cpp:110]: `desiredStopTime` in types, `desiredEndTime` here. choose one name.

File: skyrim-platform/src/platform_se/skyrim_platform/GameEventSinks.cpp [hunk 3]
```diff
@@ -665,3 +732,1297 @@ RE::BSEventNotifyControl GameEventSinks::ProcessEvent(
...
+
+  SkyrimPlatform::GetSingleton().AddUpdateTask([sleepStart, sleepEnd] {
+    auto obj = JsValue::Object();
+
+    if (sleepStart < 0 || sleepEnd < 0) {
+      return;
+    }
+
+    obj.SetProperty("startTime", JsValue::Double(sleepStart));
+    obj.SetProperty("desiredEndTime", JsValue::Double(sleepEnd));
```
Comment (Pospelove) [inline skyrim-platform/src/platform_se/skyrim_platform/GameEventSinks.cpp:1003]: `desiredStopTime` in types, `desiredEndTime` here. choose one name.

File: skyrim-platform/src/platform_se/skyrim_platform/GameEventSinks.cpp [hunk 4]
```diff
@@ -665,3 +732,1297 @@ RE::BSEventNotifyControl GameEventSinks::ProcessEvent(
...
+    return RE::BSEventNotifyControl::kContinue;
+  }
+
+  auto sleepStart = converted->sleepStartTime;
+  auto sleepEnd = converted->desiredSleepEndTime;
+
+  SkyrimPlatform::GetSingleton().AddUpdateTask([sleepStart, sleepEnd] {
+    auto obj = JsValue::Object();
+
+    if (sleepStart < 0 || sleepEnd < 0) {
```
Comment (Pospelove) [inline skyrim-platform/src/platform_se/skyrim_platform/GameEventSinks.cpp:998]: What's the meaning of this condition? Seems to be always false.
Please add a comment if this check is needed or remove.

File: skyrim-platform/src/platform_se/skyrim_platform/GameEventSinks.cpp [hunk 5]
```diff
@@ -665,3 +732,1297 @@ RE::BSEventNotifyControl GameEventSinks::ProcessEvent(
...
+  auto questId = converted->questId;
+  auto stage = converted->stage;
+
+  SkyrimPlatform::GetSingleton().AddUpdateTask([questId, stage] {
+    auto obj = JsValue::Object();
+
+    auto questLocal = RE::TESForm::LookupByID(questId);
+
+    if (questLocal == nullptr || questLocal->formType != RE::FormType::Quest ||
+        stage < 0) {
```
Comment (Pospelove) [inline skyrim-platform/src/platform_se/skyrim_platform/GameEventSinks.cpp:858]: Please remove `stage < 0` check or explain why is it needed in comment.

File: skyrim-platform/src/platform_se/skyrim_platform/GameEventSinks.cpp [hunk 6]
```diff
@@ -665,3 +732,1297 @@ RE::BSEventNotifyControl GameEventSinks::ProcessEvent(
...
+    [actor, actorId, oldLoc, oldLocId, newLoc, newLocId] {
+      auto obj = JsValue::Object();
+
+      auto actorLocal = RE::TESForm::LookupByID(actorId);
+      auto oldLocLocal = RE::TESForm::LookupByID(oldLocId);
+      auto newLocLocal = RE::TESForm::LookupByID(newLocId);
+
+      if (actorLocal == nullptr || oldLocLocal == nullptr ||
+          newLocLocal == nullptr ||
+          actorLocal->formType != RE::FormType::ActorCharacter ||
```
Comment (Pospelove) [inline skyrim-platform/src/platform_se/skyrim_platform/GameEventSinks.cpp:1063]: No need to check form types here since we already check `actorLocal != actor || oldLocLocal != oldLoc ||
          newLocLocal != newLoc`

File: skyrim-platform/src/platform_se/skyrim_platform/GameEventSinks.cpp [hunk 7]
```diff
@@ -665,3 +732,1297 @@ RE::BSEventNotifyControl GameEventSinks::ProcessEvent(
...
+  auto cell = converted->cell.get();
+  auto cellId = cell ? cell->formID : 0;
+
+  SkyrimPlatform::GetSingleton().AddUpdateTask([cell, cellId] {
+    auto obj = JsValue::Object();
+
+    auto cellLocal = RE::TESForm::LookupByID(cellId);
+
+    if (cellLocal == nullptr || cellLocal != cell ||
+        cellLocal->formType != RE::FormType::Cell) {
```
Comment (Pospelove) [inline skyrim-platform/src/platform_se/skyrim_platform/GameEventSinks.cpp:1167]: same

File: skyrim-platform/src/platform_se/skyrim_platform/GameEventSinks.cpp [hunk 8]
```diff
@@ -665,3 +732,1297 @@ RE::BSEventNotifyControl GameEventSinks::ProcessEvent(
...
+                                                casterId, spellId, status] {
+    auto obj = JsValue::Object();
+
+    auto targetLocal = RE::TESForm::LookupByID(targetId);
+    auto casterLocal = RE::TESForm::LookupByID(casterId);
+    auto spellLocal = RE::TESForm::LookupByID(spellId);
+
+    if (targetLocal == nullptr || casterLocal == nullptr ||
+        spellLocal == nullptr || targetLocal != target ||
+        casterLocal != caster || spellLocal->formType != RE::FormType::Spell ||
```
Comment (Pospelove) [inline skyrim-platform/src/platform_se/skyrim_platform/GameEventSinks.cpp:1249]: same for this formType check

File: skyrim-platform/src/platform_se/skyrim_platform/GameEventSinks.cpp [hunk 9]
```diff
@@ -665,3 +732,1297 @@ RE::BSEventNotifyControl GameEventSinks::ProcessEvent(
...
+    obj.SetProperty("target", CreateObject("ObjectReference", targetLocal));
+    obj.SetProperty("caster", CreateObject("ObjectReference", casterLocal));
+    obj.SetProperty("spell", CreateObject("Spell", spellLocal));
+    obj.SetProperty("isFriendly", JsValue::Bool(false));
+    obj.SetProperty("isWardAbsorbed", JsValue::Bool(false));
+    obj.SetProperty("isWardBroken", JsValue::Bool(false));
+
+    switch (status) {
+      case TESEvents::TESMagicWardHitEvent::Status::kFriendly: {
+        obj.SetProperty("isFriendly", JsValue::Bool(true));
```
Comment (Pospelove) [inline skyrim-platform/src/platform_se/skyrim_platform/GameEventSinks.cpp:1263]: SetProperty("status", "friendly")

File: skyrim-platform/src/platform_se/skyrim_platform/GameEventSinks.cpp [hunk 10]
```diff
@@ -665,3 +732,1297 @@ RE::BSEventNotifyControl GameEventSinks::ProcessEvent(
...
+        break;
+      }
+      case TESEvents::TESMagicWardHitEvent::Status::kAbsorbed: {
+        obj.SetProperty("isWardAbsorbed", JsValue::Bool(true));
```
Comment (Pospelove) [inline skyrim-platform/src/platform_se/skyrim_platform/GameEventSinks.cpp:1267]: SetProperty("status", "absorbed")

File: skyrim-platform/src/platform_se/skyrim_platform/GameEventSinks.cpp [hunk 11]
```diff
@@ -665,3 +732,1297 @@ RE::BSEventNotifyControl GameEventSinks::ProcessEvent(
...
+        break;
+      }
+      case TESEvents::TESMagicWardHitEvent::Status::kBroken: {
+        obj.SetProperty("isWardBroken", JsValue::Bool(true));
```
Comment (Pospelove) [inline skyrim-platform/src/platform_se/skyrim_platform/GameEventSinks.cpp:1271]: SetProperty("status", "broken")

File: skyrim-platform/src/platform_se/codegen/convert-files/Definitions.txt [hunk 12]
```diff
@@ -264,6 +264,187 @@ export interface SpellCastEvent {
...
+export interface FurnitureEvent {
+    target: ObjectReference,
+    actor: ObjectReference
+}
+
+export interface WardHitEvent {
+    target: ObjectReference,
+    caster: ObjectReference,
+    spell: Spell,
+    isFriendly: boolean,
```
Comment (Pospelove) [inline skyrim-platform/src/platform_se/codegen/convert-files/Definitions.txt:336]: better to have this:

```
status: 'friendly' | 'absorbed' | 'broken'
```

File: skyrim-platform/src/platform_se/skyrim_platform/GameEventSinks.cpp [hunk 13]
```diff
@@ -665,3 +732,1297 @@ RE::BSEventNotifyControl GameEventSinks::ProcessEvent(
...
+  SkyrimPlatform::GetSingleton().AddUpdateTask(
+    [actor, actorId, packageId, type] {
+      auto obj = JsValue::Object();
+
+      auto actorLocal = RE::TESForm::LookupByID(actorId);
+      auto packageLocal = RE::TESForm::LookupByID(packageId);
+
+      if (actorLocal == nullptr || packageLocal == nullptr ||
+          actorLocal != actor ||
+          packageLocal->formType != RE::FormType::Package ||
```
Comment (Pospelove) [inline skyrim-platform/src/platform_se/skyrim_platform/GameEventSinks.cpp:1306]: same for this type check

File: skyrim-platform/src/platform_se/skyrim_platform/GameEventSinks.cpp [hunk 14]
```diff
@@ -665,3 +732,1297 @@ RE::BSEventNotifyControl GameEventSinks::ProcessEvent(
...
+  auto actor = converted->actor.get();
+  auto actorId = actor ? actor->formID : 0;
+
+  SkyrimPlatform::GetSingleton().AddUpdateTask([actor, actorId] {
+    auto obj = JsValue::Object();
+
+    auto actorLocal = RE::TESForm::LookupByID(actorId);
+
+    if (actorLocal == nullptr || actorLocal != actor ||
+        actorLocal->formType != RE::FormType::ActorCharacter) {
```
Comment (Pospelove) [inline skyrim-platform/src/platform_se/skyrim_platform/GameEventSinks.cpp:1353]: same

File: skyrim-platform/src/platform_se/skyrim_platform/GameEventSinks.cpp [hunk 15]
```diff
@@ -665,3 +732,1297 @@ RE::BSEventNotifyControl GameEventSinks::ProcessEvent(
...
+
+  auto newState = event->newState;
+  auto newStateId = newState ? newState->id : 0;
+  auto oldState = event->oldState;
+  auto oldStateId = oldState ? oldState->id : 0;
+
+  SkyrimPlatform::GetSingleton().AddUpdateTask([oldStateId, newStateId] {
+    auto obj = JsValue::Object();
+
+    if (oldStateId < 0 || newStateId < 0) {
```
Comment (Pospelove) [inline skyrim-platform/src/platform_se/skyrim_platform/GameEventSinks.cpp:1469]: please remove this check or explain why is it needed.

File: skyrim-platform/src/platform_se/skyrim_platform/GameEventSinks.cpp [hunk 16]
```diff
@@ -665,3 +732,1297 @@ RE::BSEventNotifyControl GameEventSinks::ProcessEvent(
...
+RE::BSEventNotifyControl GameEventSinks::ProcessEvent(
+  const SKSE::CameraEvent* event,
+  RE::BSTEventSource<SKSE::CameraEvent>* eventSource)
+{
+
+  if (event == nullptr) {
+    return RE::BSEventNotifyControl::kContinue;
+  }
+
+  auto newState = event->newState;
```
Comment (Pospelove) [inline skyrim-platform/src/platform_se/skyrim_platform/GameEventSinks.cpp:1461]: we better return if newState/oldState is nullptr

File: skyrim-platform/src/platform_se/skyrim_platform/GameEventSinks.cpp [hunk 17]
```diff
@@ -665,3 +732,1297 @@ RE::BSEventNotifyControl GameEventSinks::ProcessEvent(
...
+
+RE::BSEventNotifyControl GameEventSinks::ProcessEvent(
+  const SKSE::ModCallbackEvent* event,
+  RE::BSTEventSource<SKSE::ModCallbackEvent>* eventSource)
+{
+  if (event == nullptr) {
+    return RE::BSEventNotifyControl::kContinue;
+  }
+
+  auto eventName = event->eventName;
```
Comment (Pospelove) [inline skyrim-platform/src/platform_se/skyrim_platform/GameEventSinks.cpp:1552]: it will look safer if we cast to `std::string` at this stage.

File: skyrim-platform/src/platform_se/skyrim_platform/GameEventSinks.cpp [hunk 18]
```diff
@@ -665,3 +732,1297 @@ RE::BSEventNotifyControl GameEventSinks::ProcessEvent(
...
+                      JsValue::String(static_cast<std::string>(strArg)));
+      obj.SetProperty("numArg", JsValue::Double(numArg));
+
+      EventsApi::SendEvent("modEvent", { JsValue::Undefined(), obj });
+    });
+
+  return RE::BSEventNotifyControl::kContinue;
+}
+
+RE::BSEventNotifyControl GameEventSinks::ProcessEvent(
```
Comment (Pospelove) [inline skyrim-platform/src/platform_se/skyrim_platform/GameEventSinks.cpp:1581]: These two events look strange. I think it needs more in-game testing and can be landed in a separate PR.
Please remove from this PR for now.

File: skyrim-platform/src/platform_se/skyrim_platform/GameEventSinks.cpp [hunk 19]
```diff
@@ -665,3 +732,1297 @@ RE::BSEventNotifyControl GameEventSinks::ProcessEvent(
...
+      }
+      if (enchantingFlag) {
+        EventsApi::SendEvent("itemEnchanted", { JsValue::Undefined(), obj });
+      }
+    });
+
+  return RE::BSEventNotifyControl::kContinue;
+}
+
+RE::BSEventNotifyControl GameEventSinks::ProcessEvent(
```
Comment (Pospelove) [inline skyrim-platform/src/platform_se/skyrim_platform/GameEventSinks.cpp:1617]: Same. I don't like the current interface. We better discuss this item in a separate PR.

File: skyrim-platform/src/platform_se/skyrim_platform/GameEventSinks.cpp [hunk 20]
```diff
@@ -665,3 +732,1297 @@ RE::BSEventNotifyControl GameEventSinks::ProcessEvent(
...
+      obj.SetProperty("prevCount", JsValue::Double(prevCount));
+      obj.SetProperty("newCount", JsValue::Double(newCount));
+
+      EventsApi::SendEvent("inventory", { JsValue::Undefined(), obj });
+    });
+
+  return RE::BSEventNotifyControl::kContinue;
+}
+
+RE::BSEventNotifyControl GameEventSinks::ProcessEvent(
```
Comment (Pospelove) [inline skyrim-platform/src/platform_se/skyrim_platform/GameEventSinks.cpp:1673]: No void events, please. Remove for now.

File: skyrim-platform/src/platform_se/skyrim_platform/GameEventSinks.cpp [hunk 21]
```diff
@@ -665,3 +732,1297 @@ RE::BSEventNotifyControl GameEventSinks::ProcessEvent(
...
+  SkyrimPlatform::GetSingleton().AddUpdateTask([] {
+    auto obj = JsValue::Object();
+
+    EventsApi::SendEvent("chestLooted", { JsValue::Undefined(), obj });
+  });
+
+  return RE::BSEventNotifyControl::kContinue;
+}
+
+RE::BSEventNotifyControl GameEventSinks::ProcessEvent(
```
Comment (Pospelove) [inline skyrim-platform/src/platform_se/skyrim_platform/GameEventSinks.cpp:1690]: this also cannot be merged until we will be able to access pickpocketed actor and item. remove from this PR for now.

File: skyrim-platform/src/platform_se/skyrim_platform/GameEventSinks.cpp [hunk 22]
```diff
@@ -665,3 +732,1297 @@ RE::BSEventNotifyControl GameEventSinks::ProcessEvent(
...
+    [ref, refId, sceneId, refAliasId, questId, action] {
+      auto obj = JsValue::Object();
+
+      auto refLocal = RE::TESForm::LookupByID(refId);
+      auto sceneLocal = RE::TESForm::LookupByID(sceneId);
+      auto questLocal = RE::TESForm::LookupByID(questId);
+
+      if (refLocal == nullptr || sceneLocal == nullptr ||
+          questLocal == nullptr || refAliasId < 0 || action < 0 ||
+          sceneLocal->formType != RE::FormType::Scene ||
```
Comment (Pospelove) [inline skyrim-platform/src/platform_se/skyrim_platform/GameEventSinks.cpp:1780]: check equality to original values instead of checking form type.

File: skyrim-platform/src/platform_se/skyrim_platform/GameEventSinks.cpp [hunk 23]
```diff
@@ -665,3 +732,1297 @@ RE::BSEventNotifyControl GameEventSinks::ProcessEvent(
...
+  SkyrimPlatform::GetSingleton().AddUpdateTask(
+    [weaponId, ammoId, power, isSunGazing] {
+      auto obj = JsValue::Object();
+
+      auto weaponLocal = RE::TESForm::LookupByID(weaponId);
+      auto ammoLocal = RE::TESForm::LookupByID(ammoId);
+
+      if (weaponLocal == nullptr || ammoLocal == nullptr ||
+          weaponLocal->formType != RE::FormType::Weapon ||
+          ammoLocal->formType != RE::FormType::Ammo) {
```
Comment (Pospelove) [inline skyrim-platform/src/platform_se/skyrim_platform/GameEventSinks.cpp:1822]: check equality to original values instead of checking form type.

File: skyrim-platform/src/platform_se/skyrim_platform/GameEventSinks.cpp [hunk 24]
```diff
@@ -665,3 +732,1297 @@ RE::BSEventNotifyControl GameEventSinks::ProcessEvent(
...
+  if (!converted) {
+    return RE::BSEventNotifyControl::kContinue;
+  }
+
+  auto travelTime = converted->travelTime;
+
+  SkyrimPlatform::GetSingleton().AddUpdateTask([travelTime] {
+    auto obj = JsValue::Object();
+
+    if (travelTime < 0) {
```
Comment (Pospelove) [inline skyrim-platform/src/platform_se/skyrim_platform/GameEventSinks.cpp:1853]: explain why is this check needed or remove

File: skyrim-platform/src/platform_se/skyrim_platform/GameEventSinks.cpp [hunk 25]
```diff
@@ -665,3 +732,1297 @@ RE::BSEventNotifyControl GameEventSinks::ProcessEvent(
...
+                                                targetId, perkId, flag] {
+    auto obj = JsValue::Object();
+
+    auto causeLocal = RE::TESForm::LookupByID(causeId);
+    auto targetLocal = RE::TESForm::LookupByID(targetId);
+    auto perkLocal = RE::TESForm::LookupByID(perkId);
+
+    if (causeLocal == nullptr || targetLocal == nullptr ||
+        perkLocal == nullptr || causeLocal != cause || targetLocal != target ||
+        perkLocal->formType != RE::FormType::Perk || flag < 0) {
```
Comment (Pospelove) [inline skyrim-platform/src/platform_se/skyrim_platform/GameEventSinks.cpp:1942]: check perk equality instead of form type

File: skyrim-platform/src/platform_se/skyrim_platform/GameEventSinks.cpp [hunk 26]
```diff
@@ -665,3 +732,1297 @@ RE::BSEventNotifyControl GameEventSinks::ProcessEvent(
...
+
+RE::BSEventNotifyControl GameEventSinks::ProcessEvent(
+  const RE::BGSFootstepEvent* event,
+  RE::BSTEventSource<RE::BGSFootstepEvent>* eventSource)
+{
+  if (event == nullptr) {
+    return RE::BSEventNotifyControl::kContinue;
+  }
+
+  auto tag = event->tag;
```
Comment (Pospelove) [inline skyrim-platform/src/platform_se/skyrim_platform/GameEventSinks.cpp:1986]: better to cast to string at this stage

File: skyrim-platform/src/platform_se/skyrim_platform/GameEventSinks.cpp [hunk 27]
```diff
@@ -665,3 +732,1297 @@ RE::BSEventNotifyControl GameEventSinks::ProcessEvent(
...
+
+  auto flag = converted->flag;
+  auto sceneId = converted->sceneId;
+
+  SkyrimPlatform::GetSingleton().AddUpdateTask([flag, sceneId] {
+    auto obj = JsValue::Object();
+
+    auto sceneLocal = RE::TESForm::LookupByID(sceneId);
+
+    if (sceneLocal == nullptr) {
```
Comment (Pospelove) [inline skyrim-platform/src/platform_se/skyrim_platform/GameEventSinks.cpp:2017]: also check sceneLocal == scene

**Review by Pospelove — CHANGES_REQUESTED** [review]

File: skyrim-platform/src/platform_se/codegen/convert-files/Definitions.txt [hunk 28]
```diff
@@ -264,6 +264,187 @@ export interface SpellCastEvent {
...
+
+export interface SceneActionEvent {
+    reference: ObjectReference,
+    referenceAliasId: number,
+    scene: Scene,
+    quest: Quest,
+    action: number
+}
+
+export interface SceneEvent {
```
[the remark this answers is not in the document]
Comment (shevernitskiy) [reply skyrim-platform/src/platform_se/codegen/convert-files/Definitions.txt:364]: I don't know.

File: skyrim-platform/src/platform_se/codegen/convert-files/Definitions.txt [hunk 29]
```diff
@@ -264,6 +264,187 @@ export interface SpellCastEvent {
...
+    scene: Scene,
+    flag: number
```
[the remark this answers is not in the document]
Comment (shevernitskiy) [reply skyrim-platform/src/platform_se/codegen/convert-files/Definitions.txt:366]: This is not finished event. First value is a pointer, but not to form (i dont know how to work with other possible entities). I decrypt only sceneId (second value).

File: skyrim-platform/src/platform_se/codegen/convert-files/Definitions.txt [hunk 30]
```diff
@@ -348,6 +529,174 @@ export declare function once(eventName: 'consoleMessage', callback: (event: Cons
...
+export declare function on(eventName: 'enterBleedout', callback: (event: EnterBleedoutEvent) => void): void;
+export declare function once(eventName: 'enterBleedout', callback: (event: EnterBleedoutEvent) => void): void;
+
+export declare function on(eventName: 'destructionStageChanged', callback: (event: DestructionStageChangedEvent) => void): void;
+export declare function once(eventName: 'destructionStageChanged', callback: (event: DestructionStageChangedEvent) => void): void;
+
+export declare function on(eventName: 'sceneAction', callback: (event: SceneActionEvent) => void): void;
+export declare function once(eventName: 'sceneAction', callback: (event: SceneActionEvent) => void): void;
+
+export declare function on(eventName: 'scene', callback: (event: SceneEvent) => void): void;
```
[the remark this answers is not in the document]
Comment (shevernitskiy) [reply skyrim-platform/src/platform_se/codegen/convert-files/Definitions.txt:613]: I don't know. It called TESSceneEvent in CommonLib, myabe there are many stages (maybe flag defines, what event triggered: start, end or so)

File: skyrim-platform/src/platform_se/skyrim_platform/GameEventSinks.cpp (hunk 22 above)
Comment (shevernitskiy) [reply skyrim-platform/src/platform_se/skyrim_platform/GameEventSinks.cpp:1780]: I should check `sceneLocal->formID == sceneId` after `auto sceneLocal = RE::TESForm::LookupByID(sceneId)` ?

Comment (shevernitskiy) [reply skyrim-platform/src/platform_se/skyrim_platform/GameEventSinks.cpp:1780]: check `refLocal == ref` never pass, it is probably not form

File: skyrim-platform/src/platform_se/skyrim_platform/GameEventSinks.cpp (hunk 27 above)
Comment (shevernitskiy) [reply skyrim-platform/src/platform_se/skyrim_platform/GameEventSinks.cpp:2017]: we haven't scene form in event, only sceneId, i'll add formId check

File: skyrim-platform/src/platform_se/skyrim_platform/GameEventSinks.cpp [hunk 31]
```diff
@@ -665,3 +727,1074 @@ RE::BSEventNotifyControl GameEventSinks::ProcessEvent(
...
+    [cause, causeId, target, targetId, perkId, flag] {
+      auto obj = JsValue::Object();
+
+      auto causeLocal = RE::TESForm::LookupByID(causeId);
+      auto targetLocal = RE::TESForm::LookupByID(targetId);
+      auto perkLocal = RE::TESForm::LookupByID(perkId);
+
+      if (causeLocal == nullptr || targetLocal == nullptr ||
+          perkLocal == nullptr || causeLocal != cause ||
+          targetLocal != target || perkLocal->formID != perkId || flag < 0) {
```
Comment (Pospelove) [inline skyrim-platform/src/platform_se/skyrim_platform/GameEventSinks.cpp:1745]: Sorry. We need FormType checks here back. My mistake.

File: skyrim-platform/src/platform_se/skyrim_platform/GameEventSinks.cpp [hunk 32]
```diff
@@ -665,3 +727,1074 @@ RE::BSEventNotifyControl GameEventSinks::ProcessEvent(
...
+
+  SkyrimPlatform::GetSingleton().AddUpdateTask(
+    [sceneId, refAliasId, questId, action] {
+      auto obj = JsValue::Object();
+
+      auto sceneLocal = RE::TESForm::LookupByID(sceneId);
+      auto questLocal = RE::TESForm::LookupByID(questId);
+
+      if (sceneLocal == nullptr || questLocal == nullptr || refAliasId < 0 ||
+          action < 0 || sceneLocal->formID != sceneId ||
```
Comment (Pospelove) [inline skyrim-platform/src/platform_se/skyrim_platform/GameEventSinks.cpp:1589]: Sorry. We need FormType checks here back. My mistake.

**Review by Pospelove — CHANGES_REQUESTED** [review]

**Review by Pospelove — APPROVED** [review]
